crates/bevy_ecs/src/system/query.rs

```diff
@@ -10,235 +10,265 @@ use crate::{
-/// ```
+/// # Performance
+///
+/// Creating a `Query` is a low-cost operation.
+/// Iterating it, on the other hand, fetches data from the world and generates items, which can have a significant computational cost.
+/// However, iteration cost of queries is reduced by the fact that they have an internal archetype cache to avoid re-computing archetype matches on each query access.
 ///
-/// ## Getting the query result for a particular entity
+/// Two systems cannot be executed in parallel if both access the same component type where at least one of the accesses is mutable.
+/// This happens unless the schedule can verify that no entity could be found in both queries.
 ///
-/// If you have an [`Entity`] ID, you can use the [`get`](Self::get) or
-/// [`get_mut`](Self::get_mut) methods to access the query result for that particular entity.
+/// Optional components increase the number of entities a query has to match against.
+/// This can hurt iteration performance, especially if the query solely consists of only optional components, since the query would iterate over each entity in the world.
 ///
-/// ## Getting a single query result
+/// The following table compares the computational complexity of the various methods and operations, where:
 ///
-/// While it's possible to get a single result from a query by using `iter.next()`, a more
-/// idiomatic approach would use the [`single`](Self::single) or [`single_mut`](Self::single_mut)
-/// methods instead. Keep in mind though that they will return a [`QuerySingleError`] if the
-/// number of query results differ from being exactly one. If that's the case, use `iter.next()`
-/// (or `iter_mut.next()`) to only get the first query result.
+/// - **n** is the number of entities that match the query,
+/// - **r** is the number of elements in a combination,
+/// - **k** is the number of involved entities in the operation,
+/// - **a** is the number of archetypes in the world,
+/// - **C** is the [binomial coefficient], used to count combinations.
+///   <sub>n</sub>C<sub>r</sub> is read as "*n* choose *r*" and is equivalent to the number of distinct unordered subsets of *r* elements that can be taken from a set of *n* elements.
+///
+/// |Query operation|Computational complexity|
+/// |:---:|:---:|
+/// |[`iter`]\([`_mut`][iter_mut])|O(n)|
+/// |[`for_each`]\([`_mut`][for_each_mut]),<br>[`par_for_each`]\([`_mut`][par_for_each_mut])|O(n)|
+/// |[`iter_many`],<br>[`many_for_each_mut`]|O(k)|
+/// |[`iter_combinations`]\([`_mut`][iter_combinations_mut])|O(<sub>n</sub>C<sub>r</sub>)|
+/// |[`get`]\([`_mut`][get_mut])|O(1)|
+/// |([`get_`][get_many])[`many`]|O(k)|
+/// |([`get_`][get_many_mut])[`many_mut`]|O(k<sup>2</sup>)|
+/// |[`single`]\([`_mut`][single_mut]),<br>[`get_single`]\([`_mut`][get_single_mut])|O(a)|
+/// |Archetype based filtering ([`With`], [`Without`], [`Or`])|O(a)|
+/// |Change detection filtering ([`Added`], [`Changed`])|O(a + n)|
+///
```

Comment:
Probably should add a note here that `for_each` may be conditionally faster to run than `iter` on heavily fragmented worlds, but also discourage it's use as it's generally not as flexible as `iter`.

Reply:
By heavily fragmented world, you mean a world with a high archetype fragmentation?

Reply:
That's consistent with my understanding of the performance characteristics, yes.

Reply:
I'm not sure that's still the case? From https://github.com/bevyengine/bevy/pull/4060 it's not obvious at least. In Cart benchmark, when everything is blackboxed, there are no difference.
I don't think that should be mentioned

Reply:
Hmm, interesting point. Let's see what @james7132 has to say about. If we decide to remove the performance remarks of `for_each`, we should also do the same at the method level.

Reply:
Given that we're still seeing fragmented iteration results that are still at issue here, I still would suggest putting it. From the benchmarks seen while testing #4800, there's still definitely a big difference (though that PR closes the gap significantly), particularly for wider queries that match a lot of archetypes.